hw/src/trezor.rs

```diff
@@ -0,0 +1,462 @@
+	}
+}
+
+/// Ledger device manager.
+pub struct Manager {
+	usb: Arc<Mutex<hidapi::HidApi>>,
+	devices: Vec<Device>,
+	closed_devices: Vec<String>,
+	key_path: KeyPath,
+}
+
+#[derive(Debug)]
+struct Device {
+	path: String,
+	info: WalletInfo,
+}
+
+/// HID Version used for the Trezor device
+enum HidVersion {
+	V1,
+	V2,
+}
+
+impl Manager {
+	/// Create a new instance.
+	pub fn new(hidapi: Arc<Mutex<hidapi::HidApi>>) -> Manager {
+		Manager {
+			usb: hidapi,
+			devices: Vec::new(),
+			closed_devices: Vec::new(),
+			key_path: KeyPath::Ethereum,
+		}
+	}
+
+	/// Re-populate device list
+	pub fn update_devices(&mut self) -> Result<usize, Error> {
+		let mut usb = self.usb.lock();
+		usb.refresh_devices();
+		let devices = usb.devices();
+		let mut new_devices = Vec::new();
+		let mut closed_devices = Vec::new();
+		let mut error = None;
+		for usb_device in devices {
+			let is_trezor = usb_device.vendor_id == TREZOR_VID;
+			let is_supported_product = TREZOR_PIDS.contains(&usb_device.product_id);
+			let is_valid = usb_device.usage_page == 0xFF00 || usb_device.interface_number == 0;
+
+			trace!(
+				"Checking device: {:?}, trezor: {:?}, prod: {:?}, valid: {:?}",
+				usb_device,
+				is_trezor,
+				is_supported_product,
+				is_valid,
+			);
+			if !is_trezor || !is_supported_product || !is_valid {
+				continue;
+			}
+			match self.read_device_info(&usb, &usb_device) {
+				Ok(device) => new_devices.push(device),
+				Err(Error::ClosedDevice(path)) => closed_devices.push(path.to_string()),
+				Err(e) => {
+					warn!("Error reading device: {:?}", e);
+					error = Some(e);
+				}
+			}
+		}
+		let count = new_devices.len();
+		trace!("Got devices: {:?}, closed: {:?}", new_devices, closed_devices);
+		self.devices = new_devices;
+		self.closed_devices = closed_devices;
+		match error {
+			Some(e) => Err(e),
+			None => Ok(count),
+		}
+	}
+
+	fn read_device_info(&self, usb: &hidapi::HidApi, dev_info: &hidapi::HidDeviceInfo) -> Result<Device, Error> {
+		let handle = self.open_path(|| usb.open_path(&dev_info.path))?;
+		let manufacturer = dev_info.manufacturer_string.clone().unwrap_or("Unknown".to_owned());
+		let name = dev_info.product_string.clone().unwrap_or("Unknown".to_owned());
+		let serial = dev_info.serial_number.clone().unwrap_or("Unknown".to_owned());
+		match self.get_address(&handle) {
+			Ok(Some(addr)) => {
+				Ok(Device {
+					path: dev_info.path.clone(),
+					info: WalletInfo {
+						name: name,
+						manufacturer: manufacturer,
+						serial: serial,
+						address: addr,
+					},
+				})
+			}
+			Ok(None) => Err(Error::ClosedDevice(dev_info.path.clone())),
+			Err(e) => Err(e),
+		}
+	}
+
+	pub fn message(&self, message_type: &TrezorMessageType, device_path: &Option<String>, message: &Option<String>) -> Result<String, Error> {
+		match *message_type {
+			TrezorMessageType::GetDevices => {
+				serde_json::to_string(&self.closed_devices).map_err(Error::SerdeError)
```

Comment:
Would really prefer to have a serializable enum return type as well, passing the strings around feels a bit hacky.
TBH since we only have two possible messages why not split it into separate methods?

Reply:
Yeah that's what I was asking above as well. If we change it to two methods, then `getLockedDevices` could be `lockedHardwareAccountsInfo` to match with the `hardwareAccountsInfo` method and a `hardwarePinMatrixAck` method (that could live under `parity_accounts` to be more protected than just getting a list of locked devices).

That would also solve the "enum problem" by removing it altogether. I ended up not doing an enum return type because the workaround for lack of `impl Trait` felt hackier than returning a string.

If you agree then I think I'll split it into two methods and remove this whole message-passing thing.

Reply:
Awesome, makes sense!